drivers/samples: flash: shell: fix issues

gmarull opened the pull request:
- The flash shell/sample shell included soc.h for no reason, remove it. Also removed a duplicate include.
- Fix hifive_unmatched issues with `jedec,spi-nor` (node could not be used)

Fixes #47576

minimal libc issues discovered here are being fixed in #47612

gmarull committed:
drivers: flash: shell: remove unused soc.h

The flash shell included soc.h for no reason, remove it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: flash_shell: remove unused soc.h

The sample included soc.h for no reason, remove it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: flash_shell: remove duplicate include

<stdlib.h> was included twice, keep the first occurence only.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
Investigating current failure (also on `main`):

`west build -b hifive_unmatched samples/drivers/flash_shell -p`

```
/home/gmarull/ws/nordic/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c: In function 'do_write_unaligned':
/home/gmarull/ws/nordic/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c:304:23: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  304 |   off_t data_offset = (off_t)buf;
      |                       ^
/home/gmarull/ws/nordic/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c:317:7: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  317 |       (const void *)data_offset,
      |       ^
[153/163] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map 
: && ccache /home/gmarull/zephyr-sdk/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/riscv/core/offsets/offsets.c.obj  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/riscv/core/libarch__riscv__core.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/spi/libdrivers__spi.a  zephyr/drivers/flash/libdrivers__flash.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/home/gmarull/zephyr-sdk/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/rv64imac/lp64/medany"  -L/home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -no-pie  -mcmodel=medany  -mabi=lp64  -march=rv64imac  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr && /home/gmarull/.local/lib/python3.9/site-packages/cmake/data/bin/cmake -E echo
/home/gmarull/zephyr-sdk/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: zephyr/drivers/flash/libdrivers__flash.a(spi_nor.c.obj):/home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/flash/spi_nor.c:1230: undefined reference to `__device_dts_ord_20'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

carlocaione commented:
@gmarull see #47576

gmarull commented:
Note: there seems to be a bug in minimal lib w.r.t `off_t` definition, discussed with @stephanosio who will submit a fix.

gmarull committed:
drivers: spi: sifive: remove CONFIG_SIFIVE_SPI_0_ROM

CONFIG_SIFIVE_SPI_0_ROM (default y) was an option to disable spi0 if
used to access SPI Flash ROM. However, its design had a problem: it
relied on instance numbers. You had to set status okay for spi0 to make
it work (incongruent with the purpose of the option itself). This patch
makes things simpler: if such SPI0 is not available, simply keep it
disabled in DT. Bindings have been updated to mention this case.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: riscv: hifive: disable spi0 and flash0 by default

When flash connected to spi0 is used as ROM Flash, it can't be
instantiated. spi0 is in fact disabled now with CONFIG_SIFIVE_SPI_0_ROM,
which defaults to y. However, the child flash device is not. This causes
issues with tests relying on `jedec,spi-nor` because they can't find a
parent device.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

carlescufi commented on lib/libc/minimal/include/sys/types.h:
```diff
@@ -33,25 +33,7 @@ typedef __SIZE_TYPE__ ssize_t;
-#else
-#error "The minimal libc library does not recognize the architecture!\n"
-#endif
+typedef long off_t;
```
`typdef long int off_t`

gmarull replied:
checkpatch doesn't like this (check previous runs)

gmarull committed:
arch: riscv: linker: do not require jedec,spi-nor to be 'okay'

It is not always possible to enable the jedec,spi-nor driver (e.g.
because the SPI controller can't be enabled). This is the case for
HiFive boards. Still, DT can be used for hardware description.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>
